### ISSUE TITLE
Connection to localhost not working in 2.1.0.1

### ISSUE THREAD

**Issue by Tornhoof** [jstedfast/MailKit#782]
**Describe the bug**
After upgrading Mailkit from 2.0.7 to 2.1.0.1, my E2E test with a localhost smtp server does not work anymore, as soon as I want to connect to ``localhost`` it throws a Socket Exception.
It works fine with ``127.0.0.1`` and/or Mailkit 2.0.7.

As it works fine with 2.0.7, I don't think it's a firewall problem, but instead a change in MailKit, 

**To Reproduce**
1. Take repro from below, create a project for .NET Core 2.1 and use with 2.1.0.1.
2. Run Repro

**Expected behavior**
No Exception

**Actual Behaviour**
Exception (sorry for the german text)
```csharp
System.Net.Internals.SocketExceptionFactory.ExtendedSocketException
  HResult=0x80004005
  Message=Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte
  Source=System.Private.CoreLib
  StackTrace:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at MailKit.Net.SocketUtils.SocketConnected(IAsyncResult ar)
   at System.Net.LazyAsyncResult.Complete(IntPtr userToken)
   at System.Net.ContextAwareResult.CompleteCallback()
   at System.Net.ContextAwareResult.<>c.<Complete>b__15_0(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Net.ContextAwareResult.Complete(IntPtr userToken)
   at System.Net.LazyAsyncResult.ProtectedInvokeCallback(Object result, IntPtr userToken)
   at System.Net.Sockets.BaseOverlappedAsyncResult.CompletionCallback(Int32 numBytes, SocketError socketError)
   at System.Net.Sockets.BaseOverlappedAsyncResult.CompletionPortCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)

```
**Workaround**
Use "127.0.0.1" instead of localhost

**Desktop (please complete the following information):**
 - OS: [e.g. iOS] Windows, Windows Server
 - Version [e.g. 22] 10, 2016

**Additional context**

Repro (with mini smtp server for testing):
<details>

```csharp

using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using MailKit.Net.Smtp;
using MailKit.Security;
using MimeKit;

namespace MailKitRepro
{
    class Program
    {
        static async Task Main(string[] args)
        {
            using (CancellationTokenSource cancellationTokenSource = new CancellationTokenSource())
            {
                var cancellationToken = cancellationTokenSource.Token;
                int port = 12345;
                using (var smtpListener = new SmtpListener())
                {
                    smtpListener.Start(port);
                    await SendEmailAsync(port, cancellationToken).ConfigureAwait(false);
                    cancellationTokenSource.Cancel();
                }
            }
            Console.WriteLine("Done");
            Console.ReadLine();

        }

        private static async Task SendEmailAsync(int port, CancellationToken cancellationToken = default)
        {
            try
            {
                var fromAddress = new MailboxAddress("Test User", "test@local.local");
                using (var smtpClient = new SmtpClient())
                {
                    // replace localhost with 127.0.0.1
                    await smtpClient.ConnectAsync("localhost", port, SecureSocketOptions.None, cancellationToken)
                        .ConfigureAwait(false);
                    var body = new TextPart {Text = "Hello World"};
                    var message = new MimeMessage
                        {From = {fromAddress}, Body = body, Subject = "Hello World", To = {fromAddress}};
                    await smtpClient.SendAsync(message, cancellationToken).ConfigureAwait(false);
                    await smtpClient.DisconnectAsync(true, cancellationToken).ConfigureAwait(false);
                }
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);
            }
        }
    }

    public class SmtpDataEventArgs : EventArgs
    {
        public string Content { get; set; }
    }

    /// <summary>
    /// Note: this server is actually fully synchronous
    /// </summary>
    public class SmtpListener : IDisposable
    {
        private CancellationTokenSource _cancellationTokenSource;
        private bool _disposed;
        private TcpListener _listener;
        private Task _runnerTask;

        public bool IsRunning { get; private set; }

        public int Port { get; private set; }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        public event EventHandler<SmtpDataEventArgs> EmailReceived;

        public void Start(int port)
        {
            try
            {
                _listener = new TcpListener(IPAddress.Loopback, port);
                _listener.Start(100);
                Port = port;
                _cancellationTokenSource = new CancellationTokenSource();
                var token = _cancellationTokenSource.Token;
                _runnerTask = Task.Run(() => AcceptTcpClients(token), token);
                IsRunning = true;
            }
            catch (Exception ex)
            {
                // we swallow the exception, the probability is fairly low for the error
                // the probability that it happens for a test where we need it is even lower.
            }
        }

        public void Stop()
        {
            if (_listener != null && IsRunning)
            {
                CatchAndSwallow(_cancellationTokenSource.Cancel);
                CatchAndSwallow(_listener.Stop);
                CatchAndSwallow(_runnerTask.GetAwaiter().GetResult);
                IsRunning = false;
            }
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
            {
                _disposed = true;
                Stop();
                _cancellationTokenSource?.Dispose();
            }
        }

        private async Task AcceptTcpClients(CancellationToken cancellationToken)
        {
            while (!cancellationToken.IsCancellationRequested)
            {
                try
                {
                    using (var socket = await _listener.AcceptSocketAsync().ConfigureAwait(false))
                    {
                        await HandleStreamAsync(socket, cancellationToken).ConfigureAwait(false);
                    }
                }
                catch (ObjectDisposedException ex)
                {
                    if (!_disposed)
                    {
                        // Swallow exception and log it                        
                    }
                }
            }
        }

        private void CatchAndSwallow(Action action)
        {
            try
            {
                action();
            }
            catch (TaskCanceledException)
            {
                // do nothing here
            }
            catch (Exception ex)
            {
                // swallow exception and log it                        
            }
        }

        /// <summary>
        /// Main functionality:
        /// HttpMethod is called when the listening socket receives data
        /// The EmailReceived event is invoked with the supplied data
        /// </summary>
        private async Task HandleStreamAsync(Socket socket, CancellationToken cancellationToken)
        {
            using (var stream = new NetworkStream(socket, true))
            {
                stream.ReadTimeout = -1;
                stream.WriteTimeout = -1;

                using (TextReader tr = new StreamReader(stream))
                {
                    using (TextWriter tw = new StreamWriter(stream))
                    {
                        await tw.WriteLineAsync("220 localhost -- Fake proxy server").ConfigureAwait(false);
                        await tw.FlushAsync().ConfigureAwait(false);
                        string line;
                        var content = new StringBuilder();
                        while (!cancellationToken.IsCancellationRequested && (line = await tr.ReadLineAsync().ConfigureAwait(false)) != null)
                        {
                            switch (line)
                            {
                                case "QUIT":
                                    {
                                        return;
                                    }
                                case "DATA":
                                    {
                                        await tw.WriteLineAsync("354 Start input, end data with <CRLF>.<CRLF>").ConfigureAwait(false);
                                        await tw.FlushAsync().ConfigureAwait(false);
                                        var lastLineEmpty = false;
                                        while ((line = await tr.ReadLineAsync().ConfigureAwait(false)) != null)
                                        {
                                            if (line == "." && lastLineEmpty)
                                            {
                                                var emailText = content.ToString();
                                                EmailReceived?.Invoke(this, new SmtpDataEventArgs { Content = emailText });
                                                content.Clear();
                                                break;
                                            }

                                            lastLineEmpty = string.IsNullOrEmpty(line);
                                            content.AppendLine(line);
                                        }

                                        await tw.WriteLineAsync("250 OK").ConfigureAwait(false);
                                        await tw.FlushAsync().ConfigureAwait(false);
                                        break;
                                    }
                                default:
                                    {
                                        await tw.WriteLineAsync("250 OK").ConfigureAwait(false);
                                        await tw.FlushAsync().ConfigureAwait(false);
                                        break;
                                    }
                            }
                        }
                    }
                }
            }
        }

        ~SmtpListener()
        {
            Dispose(false);
        }
    }
}

```
</details>

CsProject

<details>

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="MailKit" Version="2.1.0.1" />
  </ItemGroup>

</Project>

```
</details>

**Comment by jstedfast** [comment]
Thank you for this amazing bug report - it made it very simple for me to reproduce. I should have a fix shortly.

**Comment by jstedfast** [comment]
The above commit fixes the issue for me, but if you could verify the fix, that would be awesome.

I'll roll a new release as soon as you can verify.

Thanks again for your awesome bug report!

**Comment by Tornhoof** [comment]
Thank you for fixing the problem. I couldn't find any nuget build artifacts for the build above, so I build the version from the sources myself.
And I can confirm that with your changes above my repro works again and I can connect to localhost again.

Again, thank you very much.

**Comment by jstedfast** [comment]
Just released 2.1.0.2 with a fix (I actually modified the fix a bit more, but it still works for me locally - if it doesn't work for you, let me know and I'll look into it again).

**Comment by Tornhoof** [comment]
Yes, the nuget package 2.1.0.2 solves this bug. Thank you again.

**Comment by Tornhoof** [comment]
@jstedfast 
Hmm, I'm not 100% certain, but I think there are still issues in the connect logic.
I get a few cases of OperationCanceledException in some tests under load, ~5% at the moment
As that specific code never had any problems in 2.0.7 and earlier (with several thousands of successful runs), it might be some kind of Threadpool starvation, because your Task.Run needs another thread.

I admit I haven't done any larger socket work for years, but looking at your socket code for async:
https://github.com/jstedfast/MailKit/blob/9460cc49c77df647e80e48ac4211dede78503d35/MailKit/Net/SocketUtils.cs#L86-L90

this is imho the wrong approach, usually you wrap the old APM model, not via Task.Run but instead via 
`` Task.Factory.FromAsync``

In https://stackoverflow.com/questions/17432306/task-factory-startnew-vs-task-factory-fromasync there is a comparison of different techniques and the 
```csharp
var task = Task.Factory.FromAsync(BeginIOMethod, EndIOMethod, ... );
await task;
```
is the one without any extra threads.
This might only show up on systems under load or systems with few resources, atleast on my dev system with many cores I couln't get it reproduced easily (I had one failure, but that might have been a hiccup somewhere).

```csharp
 System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at MailKit.Net.SocketUtils.Wait(IAsyncResult ar, CancellationToken cancellationToken)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at MailKit.Net.SocketUtils.ConnectAsync(String host, Int32 port, IPEndPoint localEndPoint, Boolean doAsync, CancellationToken cancellationToken)
   at MailKit.Net.SocketUtils.ConnectAsync(String host, Int32 port, IPEndPoint localEndPoint, Int32 timeout, Boolean doAsync, CancellationToken cancellationToken)
   at MailKit.MailService.ConnectSocket(String host, Int32 port, Boolean doAsync, CancellationToken cancellationToken)
```

**Comment by jstedfast** [comment]
I was actually about to switch over to TaskFactory.FromAsync(), so that is probably good advice. I didn't do that originally because I didn't think it was available for .NET 4.5 (but as I discovered last night, it is).

That said, your exception is coming from the cancellationToken.

I'm assuming that you aren't actually calling cancel on any cancellation token you are passing to SmtpClient.Connect() (if you are even passing it a cancellation token).

I suspect what might be happening is that starting with MailKit 2.1, I made MailKit actually make use of the SmtpClient.Timeout value by creating a CancellationTokenSource with that timeout value which is *probably* what is firing.

By default, SmtpClient's Timeout value is 2 minutes, but you could set it higher.

I've make the Task.Factory.FromAsync change here: https://github.com/jstedfast/MailKit/commit/b188d25037c1755d0414a99fae3695e94638daf8

Could you let me know how that works for you?

**Comment by Tornhoof** [comment]
Thank you again for your quick reply.
I'll let the tests run in an while loop over night on a resource-limited system, to see if anything changes.
As the code internally is running in while(loop) in a BackgroundService in ASP.NET Core, it has a valid CancellationToken but that one is only triggered on application shutdown and later connect attempts were also successful, so it should not be the CancellationToken. I added a bit of logging around the CancellationToken to see if I missed anything.

I'll report back tomorrow morning with the result of my test.

**Comment by jstedfast** [comment]
I doubt it's your cancellationToken, like I said, I think it might be the timeout logic that I added to Connect[Async].

**Comment by Tornhoof** [comment]
I ran my tests over night with the latest bits from master. https://github.com/jstedfast/MailKit/commit/2148a57d9d0a2582028bb2f9fdfea853cacffd52 and I didn't see any errors anymore, the run over night created a few hundred iterations, so I'm fairly confident.

As for the timeout logic, I'd be surprised if the test took 2 minutes or longer to connect to a local smtp server, my original logs were not detailed enough to track those values. My extended log messages from the nightly run are extensive enough that the connection on average takes less than 100 milliseconds (the resolution of the timestamp is not higher).

The bottom line is, I still think it was some kind of thread starvation problem, from the Task.Run and your new code fixes that.

**Comment by jstedfast** [comment]
Thanks for the update. I think you are probably right about the Task.Run thread starvation causing the timeouts. I'm glad things are working now and I'll be making a new release later today with this fix.

**Comment by Beksultan1** [comment]
Hi! jstedfast, I have some problem, how to fix this error ?

Error !
{"@t":"2020-02-24T11:44:04.0086885Z","@m":"The operation has timed out.","@i":"e548ad2e","@l":"Error","@x":"System.TimeoutException: The operation has timed out.\n   at MailKit.Net.SocketUtils.ConnectAsync(String host, Int32 port, IPEndPoint localEndPoint, Int32 timeout, Boolean doAsync, CancellationToken cancellationToken)\n   at MailKit.MailService.ConnectSocket(String host, Int32 port, Boolean doAsync, CancellationToken cancellationToken)\n   at MailKit.Net.Imap.ImapClient.ConnectAsync(String host, Int32 port, SecureSocketOptions options, Boolean doAsync, CancellationToken cancellationToken)\n   at FileParser.ImapClient.ConectToImap() in /app/ImapClient.cs:line 46"}


And, connection to server.
`
            using (var client = new MailKit.Net.Imap.ImapClient(new ProtocolLogger(Console.OpenStandardError())))
            {
                try
                {
                    using (var cancel = new CancellationTokenSource())
                    {
                         client.ServerCertificateValidationCallback = (sender, cert, chain, error) => true;
                        client.Connect(_config.GetSection("ImapSettings")["HOST"], int.Parse(_config.GetSection("ImapSettings")["PORT"]), MailKit.Security.SecureSocketOptions.Auto);
                    
                        client.AuthenticationMechanisms.Remove(_config.GetSection("ImapSettings")["AUTHENTICATION_MECHANISMS_REMOVE_AUTH"]);
                        client.AuthenticationMechanisms.Remove(_config.GetSection("ImapSettings")["AUTHENTICATION_MECHANISMS_REMOVE_NTLM"]);
                        client.Authenticate(_config.GetSection("ImapSettings")["ACCOUNT_USER_NAME"], _config.GetSection("ImapSettings")["ACCOUNT_PASSWORD"], cancel.Token);

   }
}
`

**Comment by jstedfast** [comment]
@Beksultan1 Set a bigger timeout? Make sure your system can connect to the server under the specified timeout value?

The default timeout is 2 minutes: https://github.com/jstedfast/MailKit/blob/master/MailKit/Net/Imap/ImapClient.cs#L67

**Comment by Beksultan1** [comment]
@jstedfast , thank you for reply. I will test and show result tomorrow.

**Comment by Beksultan1** [comment]
@jstedfast , Hi! I set 2,3,4 minutes, same error.  

> {"@t":"2020-02-25T03:05:03.6046058Z","@m":"The operation has timed out.","@i":"e548ad2e","@l":"Error","@x":"System.TimeoutException: The operation has timed out.\n   at MailKit.Net.SocketUtils.ConnectAsync(String host, Int32 port, IPEndPoint localEndPoint, Int32 timeout, Boolean doAsync, CancellationToken cancellationToken)\n   at MailKit.MailService.ConnectSocket(String host, Int32 port, Boolean doAsync, CancellationToken cancellationToken)\n   at MailKit.Net.Imap.ImapClient.ConnectAsync(String host, Int32 port, SecureSocketOptions options, Boolean doAsync, CancellationToken cancellationToken)\n   at FileParser.ImapClient.ConectToImap() in /app/ImapClient.cs:line 44"}

**Comment by Beksultan1** [comment]
My code:

```
 public async Task<bool> ConectToImap()
        {
            using (var client = new MailKit.Net.Imap.ImapClient(new ProtocolLogger(Console.OpenStandardError())))
            {
                try
                {
                    using (var cancel = new CancellationTokenSource())
                    {
                        client.Timeout = int.Parse(_config.GetSection("ImapSettings")["TIME_OUT"]);
                        client.ServerCertificateValidationCallback = (sender, cert, chain, error) => true;
                        await client.ConnectAsync(_config.GetSection("ImapSettings")["HOST"], int.Parse(_config.GetSection("ImapSettings")["PORT"]), MailKit.Security.SecureSocketOptions.Auto);
                        // Log.Information($"Start contecting to Imap {client.ServerCertificateValidationCallback}");
                        client.AuthenticationMechanisms.Remove(_config.GetSection("ImapSettings")["AUTHENTICATION_MECHANISMS_REMOVE_AUTH"]);
                        client.AuthenticationMechanisms.Remove(_config.GetSection("ImapSettings")["AUTHENTICATION_MECHANISMS_REMOVE_NTLM"]);
                        client.Authenticate(_config.GetSection("ImapSettings")["ACCOUNT_USER_NAME"], _config.GetSection("ImapSettings")["ACCOUNT_PASSWORD"], cancel.Token);

                        if (client.IsConnected)
                        {
                            Log.Information($"Client connection is {client.IsConnected}");
                            SentrySdk.AddBreadcrumb("ImapClient is connected", category: "ConectToImap", level: BreadcrumbLevel.Debug);

                            client.Inbox.Open(FolderAccess.ReadWrite);
                            Log.Information($"Opened INBOX, folder access to read/write");
                            SentrySdk.AddBreadcrumb("Opened INBOX, and folder access read/write", category: "ConectToImap", level: BreadcrumbLevel.Debug);

                            //var msgIds = await client.Inbox.SearchAsync(SearchQuery.NotSeen.And(SearchQuery.NotAnswered.And(SearchQuery.NotDeleted)));
                            var msgIds = await client.Inbox.SearchAsync(SearchQuery.NotSeen);
                            Log.Information($"Get list of message Ids by asynchronously search the folder for messages matching the specified query");
                            SentrySdk.AddBreadcrumb("Get list of message Ids by asynchronously search the folder for messages matching the specified query", category: "ConectToImap", level: BreadcrumbLevel.Debug);

                            foreach (var msgId in msgIds)
                            {
                                // attachment - чтение прикрепленных файлов
                                var arch = client.Inbox.GetMessage(msgId).Attachments
                                                                        .Where(a => a.GetType() == typeof(MimePart)
                                                                                    &&
                                                                                    ((MimePart)a).FileName.ToLower().EndsWith("rar")
                                                                               )
                                                                        .Select(a => (MimePart)a)
                                                                        .FirstOrDefault();

                                Log.Information($"Get specifies message which file name ends with rar");
                                SentrySdk.AddBreadcrumb("Get specifies message which file name ends with rar", category: "ConectToImap", level: BreadcrumbLevel.Debug);

                                //using (var stream = ArchiverService.ExtractRarFile(arch))
                                //{
                                //    if (arch != null && _fileRow.CheckAllRow(stream))
                                //    {
                                //        stream.Seek(0, SeekOrigin.Begin);
                                //        _loader.LoadData(stream);
                                //        client.Inbox.SetFlags(msgId, MessageFlags.Seen, true);
                                //        Log.Information($"The message has been read. And MessageFlag is Seen");
                                //        SentrySdk.AddBreadcrumb("The message has been read. And MessageFlag signed to SEEN", category: "ConectToImap", level: BreadcrumbLevel.Debug);
                                //    }
                                //    else
                                //    {
                                //        client.Inbox.SetFlags(msgId, MessageFlags.Seen, true);
                                //        Log.Information($"Get specifies message which file name ends with rar");
                                //        throw new Exception("The message has been read. And message flag signed to SEEN. The attachment in INBOX not or invalid folder format rar");
                                //    }
                                //};
                            }
                        }
                    }
                }
                catch (ImapProtocolException excprot)
                {
                    Log.Error(excprot, excprot.Message);
                    SentrySdk.CaptureMessage(excprot.Message);
                }
                catch (Exception exc)
                {
                    Log.Error(exc, exc.Message);
                    SentrySdk.CaptureException(exc);
                }

                client.Disconnect(true);
                Log.Information($"Disconnect ImapClient");
                SentrySdk.AddBreadcrumb("Disconnect ImapClient", category: "ConectToImap", level: BreadcrumbLevel.Debug);
            }

            return true;
        }
```

**Comment by jstedfast** [comment]
@Beksultan1 if it works locally, then you have a networking and/or network interface problem on the machine you are deploying your app to.

There's no code changes that you or I can make to make it work.

**Comment by jstedfast** [comment]
@Beksultan1 Check that your deployment target doesn't have a firewall that blocks any outgoing connections. If the IMAP server you are trying to connect to is on a private network, make sure that your deployment machine actually has a route to the IMAP server host.

**Comment by Beksultan1** [comment]
@jstedfast, I checked by telnet. The server has access to the mail server

**Comment by jstedfast** [comment]
@Beksultan1 Then I don't know what the problem could be :-\

You will have to debug it I guess and see what the problem is...
